lib/wibox/container/conditional.lua

```diff
@@ -0,0 +1,362 @@
+            assert(type(ret) == "boolean", "The conditional layout `default_when`"..
+                " must return a boolean"
+            )
+
+            if ret then
+                selected = v
+                break
+            end
+        else
+            assert(false, "A conditional template has neither a `when` property"..
+                " nor a default_when"
+            )
+        end
+    end
+
+    -- Nothing changed, move on
+    if selected == self._private.selected then
+        if selected and selected.updated_callback then
+            selected.updated_callback(self._private.argument_callback())
+        end
+
+        return
+    end
+
+    if self._private.selected and self._private.selected.deselected_callback then
+        self._private.selected:deselected_callback()
+    end
+
+    self._private.selected = selected
+
+    if self._private.selected and self._private.selected.selected_callback then
+        self._private.selected:selected_callback()
+    end
+
+    local old_widget = self._private.widget
+
+    -- Create the widget unless it's already one.
+    self._private.widget = selected and
+        base.make_widget_from_value(selected) or nil
+
+    self:emit_signal("property::selected_template",
+        selected,
+        old_selected,
+        self._private.widget,
+        old_widget
+    )
+
+    self:emit_signal("widget::layout_changed")
+    self:emit_signal("widget::redraw_needed")
+end
+
+--- Add a signal source to automatically call `eval()`.
+--
+-- @DOC_wibox_container_conditional_object_signals_EXAMPLE@
+--
+-- @tparam gears.object object The source object.
+-- @tparam string signal The signal name.
+function conditional:attach_signal(object, signal)
+    table.insert(self._private.attached_signal, {object, signal})
+    object:connect_signal(signal, self._private.eval_handler)
+end
+
+--- Remove a signal source to automatically call `eval()`.
+function conditional:detach_signal(object, signal)
+    object:disconnect_signal(signal, self._private.eval_handler)
+    for k, v in ipairs(self._private.attached_signal) do
+        if v[1] == object and v[2] == signal then
+            table.remove(self._private.attached_signal, k)
+        end
+    end
+end
+
+-- Magical variable to avoid the template instantiation ahead of time.
+conditional._accept_templates = true
+
+--- Returns a new conditional container.
+-- @param[opt] widget The widget to display.
+-- @tparam[opt={}] table args Options for this widget.
+-- @function wibox.container.conditional
+local function new(args)
+    local ret = base.make_widget(nil, nil, {enable_properties = true})
+
+    gtable.crush(ret, conditional, true)
+
+    ret._private.exec_once = false
+    ret._private.attached_signal = {}
+
+    ret._private.eval_handler = function()
+        if not ret._private.exec_once then
+            gtimer.delayed_call(function()
+                ret:eval()
+            end)
```

Comment:
`gtimer.delayed_call(ret.eval, ret)` avoids creation of the anonymous function, but is less readable. Your call.
(And of course this optimisation is no longer possible if you were to add `exec_once = false` here.